nvflare/app_opt/confidential_computing/cc_manager.py

```diff
@@ -11,158 +11,240 @@
+
+        if not isinstance(verify_frequency, int):
+            raise ValueError(f"verify_frequency must be in, but got {verify_frequency.__class__}")
+        self.verify_frequency = int(verify_frequency)
+
+        self.critical_level = critical_level
+        if self.critical_level not in [SHUTDOWN_SYSTEM, SHUTDOWN_JOB]:
+            raise ValueError(f"critical_level must be in [{SHUTDOWN_SYSTEM}, {SHUTDOWN_JOB}]. But got {critical_level}")
+
+        self.verify_time = None
+        self.cc_issuers = {}
+        self.cc_verifiers = {}
         self.participant_cc_info = {}  # used by the Server to keep tokens of all clients
 
+        self.token_submitted = False
+        self.lock = threading.Lock()
+
     def handle_event(self, event_type: str, fl_ctx: FLContext):
         if event_type == EventType.SYSTEM_BOOTSTRAP:
             try:
-                err = self._prepare_for_attestation(fl_ctx)
+                self._setup_cc_authorizers(fl_ctx)
+
+                err = self._generate_tokens(fl_ctx)
             except:
                 self.log_exception(fl_ctx, "exception in attestation preparation")
                 err = "exception in attestation preparation"
             finally:
                 if err:
                     self.log_critical(fl_ctx, err, fire_event=False)
                     raise UnsafeComponentError(err)
-        elif event_type == EventType.BEFORE_CLIENT_REGISTER:
+        elif event_type == EventType.BEFORE_CLIENT_REGISTER or event_type == EventType.BEFORE_CLIENT_HEARTBEAT:
             # On client side
-            self._prepare_token_for_login(fl_ctx)
-        elif event_type == EventType.CLIENT_REGISTERED:
+            self._prepare_cc_info(fl_ctx)
+        elif event_type == EventType.CLIENT_REGISTERED or event_type == EventType.AFTER_CLIENT_HEARTBEAT:
             # Server side
             self._add_client_token(fl_ctx)
-        elif event_type == EventType.AUTHORIZE_COMMAND_CHECK:
-            command_to_check = fl_ctx.get_prop(key=FLContextKey.COMMAND_NAME)
-            self.logger.debug(f"Received {command_to_check=}")
-            if command_to_check == AdminCommandNames.CHECK_RESOURCES:
-                try:
-                    err = self._client_to_check_participant_token(fl_ctx)
-                except:
-                    self.log_exception(fl_ctx, "exception in validating participants")
-                    err = "Participants unable to meet client CC requirements"
-                finally:
-                    if err:
-                        self._not_authorize_job(err, fl_ctx)
-        elif event_type == EventType.BEFORE_CHECK_CLIENT_RESOURCES:
+        elif event_type == EventType.CLIENT_QUIT:
             # Server side
+            self._remove_client_token(fl_ctx)
+        elif event_type == EventType.BEFORE_CHECK_RESOURCE_MANAGER:
+            # Client side: check resources before job scheduled
+            try:
+                err = self._client_to_check_participant_token(fl_ctx)
```

Comment:
I think to be able to use all common functionality is good.

Should we introduce two new classes for the different functionalities and the common functionality remains here so both can either inherit or reuse?

Reply:
The current CCmanager is designed to be used in both server and client side. We could split them into two different classes. But there will be a lot of common functions. I don't see the huge benefit with this yet. But we may see the with more functions in the future.

Reply:
I am suggesting the common functions stays in this CCManager class.

And then we create ServerCCManager (with methods needed / used on server side) that will only be used at server.
And then we create ClientCCManager (with methods needed / used on client side) that will only be used at client.

But that is just a format change, not logic change so not blocking this PR.

Reply:
Leave this future consideration.